### ISSUE TITLE
Is aria-flowto essential for Graphics ARIA?

### ISSUE THREAD

**Issue by joanmarie** [w3c/graphics-aria#1]
There is some discussion in an ARIA spec issue about the possibility of deprecating `aria-flowto`: https://github.com/w3c/aria/issues/630. And `aria-flowto` is used in the examples of the Graphics ARIA spec. Given that the ARIA Working Group would like to see Graphics ARIA and Graphics AAM advance to CR in the near future:

* If `aria-flowto` is essential to Graphics ARIA, chiming in on the ARIA issue with specifics would be helpful.
* If `aria-flowto` is not essential to Graphics ARIA, can we remove its usage from the examples so that we don't wind up shipping a spec around the time we deprecate something in its contents?

**Comment by joanmarie** [comment]
@AmeliaBR: What say you? :) And thanks in advance!

**Comment by LJWatson** [comment]
Navigation within SVG is a thing yet to be solved. Given the almost complete lack of support for `-flowsto` though, I don't think we should be including it in the Graphics module examples in any case.

**Comment by AmeliaBR** [comment]
The [electrical circuit example](https://github.com/w3c/graphics-aria/blob/master/img/circuit-diagram.svg) used in the spec still makes sense without `aria-flowto`; if the general mood is that `aria-flowto` is not best practice, then we can remove it.  I could maybe add descriptions to the wire paths explicitly stating the "is connected to" relationships.

As @LJWatson notes, the issue of describing document flow and connections is a particular problem with graphics, so if the decision is to dump `aria-flowto`, I hope it's part of a larger plan to come up with a better solution.

**Comment by medeaoblongata** [comment]
I **desperately** need `aria-flowto` to be supported (and fully fleshed out, with sufficient techniques for **choosing** from amongst multiple ids). I need it to work in conjunction with the Graphics module, especially in SVG, but also in HTML.

We are **struggling** to make flowcharts accessible. We have dozens of them. They are currently in SVG. A linear text description is not the same thing at all. We already have that in the textbook, and a duplicate text description, with every 'false' branch interleaved into the prose becomes cognitive noise. 

I should point out that our algorithm flowcharts are for training acute life saving procedures. We want our users to be able to think and act **fast** before someone dies. We want our users to consult the flowchart as mnemonic aid as they run a simulated scenario. Plowing through long texts and irrelevant (possibly harmful) steps is not an acceptable solution for this use case.

WCAG's requirement that 'if there is more than one meaningful sequence, only one sequence need be represented' is **nonsense** for making flowcharts accessible. A flowchart will **never** be meaningfully represented if some branches are entirely (and arbitrarily) omitted from the accessibility tree. If I provide an example 'choice box' from one of our flowcharts: "Is the victim breathing?" with branches "YES" and "NO", you will appreciate that choosing the correct branch is more than simply meaningful. It is potentially **life critical**.

Reductio ad absurdum. By the letter-of-the-law, I can make a flowchart WCAG-compliant today simply by omitting branches, and not allowing the user to choose. Frankly, that law is an ass. If the branch you need to take happens to be one of the omitted branches, **there is no meaningful sequence**. I hope we can do better, and `aria-flowto` comes closest to a solution today. So near, and yet so far.

From my point of view, `aria-flowto` would be essential if it weren't so crippled by the lack of specification for a choosing mechanism. PLEASE do not deprecate it without a properly considered replacement. The current spec -as far as it goes- is **not wrong**. It's just **incomplete** enough to make it useless. Why can't we give `aria-flowto` some love and let it solve actual, known problems (currently unhandled) such as browsable electrical diagrams, flowcharts, organisational charts, directed graphs etc.?

I am extremely keen to correspond on this topic, and would love to share some of my experiments. Feel free to PM me.

**Comment by LJWatson** [comment]
Would a keyboard supported tree pattern work in the case of your flowchart?

As a user I have to admit if I needed to use something with aria-flowto 
on it, I'd have to look up my screen reader's shortcuts before I could. 
When time is of the essence as you mention, that would likely be a problem.

**Comment by medeaoblongata** [comment]
The tree pattern seems a poor fit. Citation from WAI-ARIA practices:

> "items that have children may be expanded or collapsed to show or hide the children"

I don't need this behavior, and I don't want to mislead users into expecting that the boxes will expand and collapse to show and hide children. I don't need the 'leaves' of the tree to perform any action, or to be 'selectable' (a substantial part of the Tree pattern).

Apart from choosing branches to browse, I don't really need 'operability' at all, and I would prefer not to use tab to navigate (since it implies operability), I just need a very simple way to steer the browse flow when there is a choice. And I want to be able to have browsable rich text (headings, lists etc.) inside the boxes. Any 'operable' widget (such as a tree) will thus require each box to have a document role!

In short, I need the promises of aria-flowto to be fulfilled.

The tree **grid** pattern has similar problems for browsable flowchart representation, and it would impose an arbitrary spatial grid on something which is more properly a topological structure. It is also antagonistic to truly responsive layouts. I would like to be able to layout the flowchart on multiple screen sizes and orientations, with the same topology. Why should a non-sighted user need to know where (spatially) the preferred box is placed on screen? The important thing is not "left, right, up, down" but "next", "previous" and "choice".

Your comment about keyboard instruction is noted. We have added an aria description to the flowcharts which indicates the (current) keyboard functionality.

**Comment by LJWatson** [comment]
>Why should a non-sighted user need to know where (spatially) the 
preferred box is placed on screen?>

Well, speaking for myself, I like to know the visual composition of 
content, perhaps because I was once sighted and that information still 
has meaning to me.

 >The important thing is not "left, right, up, down" but "next", 
"previous" and "choice".>

The pattern of mapping previous/next to left/right has been around for a 
very long time. It's used whether or not left/right makes literal sense 
or not - the next track on a media player is not actually on the left, 
yet the little arrows on the next button will often point that way.

That said, not all screen reader users are blind, so mapping the keys to 
the appropriate visual direction would make sense, even if aria-flowto 
wasn't so badly supported and completely unfamiliar to most users.

**Comment by medeaoblongata** [comment]
I have no issue with using arrow keys. I'd be happy to use "Down" for next and "Up" for previous, with "Left" and "Right" to choose branches where there is a choice. All good.

My objection is to the idea that pressing "Left" should move focus to the box to the left of the current box. That's just not going to work. The problem is more that the 'next' box in the sequence is very often not spatially adjacent to the one you are currently 'on'. The 'next' box in the sequence might be in the opposite corner of the flowchart in relation to the current box, with a long kinked arrow leading there. 

Additionally, I want to constrain users from reading the boxes in the wrong order. i.e. I want to enforce a meaningful sequence (or more often, sequences), rather than offer free movement about some notional grid.

I am attaching a PDF of one of the (many) algorithm flowcharts we are trying to turn into accessible figures. This PDF comes from existing e-learning materials, and has been auto-tagged by Acrobat, as part of a minimal remediation effort, but the tags do not indicate which connecting arrows connect to which boxes. The 'read order' is also not the same as the box sequence, not least because there are multiple branches, including loops, and each of those sequences is meaningful. This latter problem is the main thing I hope to address, although I expect to use SVG rather than PDF. 

Even if non-sighted users will miss some details, maybe this PDF will help illustrate the problem to a few more folks here better than my prose does.

Please notice the non-adjacent placement of various box sequences, and the rich text. If someone can tell me how to make this kind of content accessible and WCAG compliant using the existing ARIA spec I am very eager to be guided. I would especially appreciate any suggestions about whether we should regard such content as 'operable' or 'browsable', which roles would be most suitable to express it, and whether the ARIA spec needs additional features to support the most correct implementation. 

If we absolutely must drop aria-flowto, what replacement will meet its promises, and where is that replacement being discussed today?

[algorithm.pdf](https://github.com/w3c/graphics-aria/files/4165151/algorithm.pdf)
